Reload csv on update

flothi opened the pull request:
Dieser PR ist (noch) unvollständig, anzupassen:
- csv-dir abhängig von der Installation (kleinigkeit)
- BOSWatch beendet nicht sauber, da der Thread offensichtlich hängt (Großigkeit...)

Betrifft #261 

Beschreibung des commit:
- create a function to watch for changes in csv-dir
- import pyinotify (use pip to install it if necessary) and threading
- starting thread to work parallel
- end thread while exiting boswatch

flothi committed:
Update boswatch.py

- create a function to watch for changes in csv-dir
- import pyinotify (use pip to install it if necessary) and threading
- starting thread to work parallel
- end thread while exiting boswatch

flothi commented:
Noch kurz als Erklärung, wie ich mir das ganze gedacht (und zum Teil mit Code-Schnippseln zusammengeklaut) habe:

Die Funktion csv_watch ist das Arbeitstier, wenn es um die Beobachtung des Verzeichnis geht. Der watchmanager (wm) kriegt übergeben, dass er bei einer Änderung der Dateien im Verzeichnis oder beim Erstellen einer Datei eine Aktion ausführen soll, hier das Nachladen der Beschreibungen.

Hinweise zu pyinotify sind hier: https://github.com/seb-m/pyinotify

Prinzipiell klappt das ganze auch; der Thread wird gestartet, die Aktualisierungen werden eingeladen und beim nächsten Auftreten des Alarms auch ausgegeben.

Was nicht klappt, ist das Beenden von Boswatch - das könnte sein, dass es _eigentlich_ keine Möglichkeit gibt, einen Thread zu stoppen (daher der Kunstgriff mit der Variable run, gefunden auf sourceforge).

Eventuell hat jemand noch ne Idee, wie man das anders angehen kann.

Schrolli91 commented:
https://docs.python.org/2/library/threading.html#thread-objects
Siehe Daemon Threads, villt hilft das

speziell hier:
https://docs.python.org/2/library/threading.html#threading.Thread.daemon
`The entire Python program exits when no alive non-daemon threads are left.`

Sprich wenn du den Thread als Daemon startest, dann stirbt der mit, wenn der main Thread beendet wird.

flothi committed:
Update boswatch.py

- daemonize observing thread
- csv-dir depending on installation-dir
- cleaning up the code

flothi commented:
Super Hinweis, danke - was so ein einzelnes Flag doch auslösen kann.

Habs eingebaut, und - oh Wunder - es tut was es soll ;-)

Schrolli91 commented:
@flothi 
Gabs hier inzwischen irgendwelche Probleme? Oder tut das ganze was es soll?
Denn wenn das hier alles passt, würde ich so langsam mergen.

flothi commented:
Probleme gab's keine, zumindest sind mir keine untergekommen

Schrolli91 commented:
Wenn ich das Richtig sehe, wird ein neuer Thread gestartet, indem die CSV Files überwacht und bei Bedarf neu nachgeladen werden?

Wird der Thread immer gestartet? Kann man das deaktivieren? Evtl macht es sinn diese Funktion in der config zu hinterlegen. Wer keine CSV Files nutzt braucht auch kein dynamisches nachladen und kann sich demzufolge die Prozessorlast auch einsparen,

flothi commented:
Der wird aktuell immer gestartet, man könnte das sicherlich schaltbar machen

Da fehlt mir allerdings aktuell die Zeit für

Schrolli91 committed:
Merge branch 'develop' into pyinotify
